tests/utils_tests/test_functional.py

```diff
@@ -47,43 +49,107 @@ def __bytes__(self):
         self.assertEqual(str(t), "Î am ā Ǩlâzz.")
         self.assertEqual(bytes(t), b"\xc3\x8e am \xc4\x81 binary \xc7\xa8l\xc3\xa2zz.")
 
+    def assertCachedPropertyWorks(self, attr, Class):
+        with self.subTest(attr=attr):
+            def get(source):
+                return getattr(source, attr)
+
+            # docstring should be preserved
+            self.assertEqual(get(Class).__doc__, 'Here is the docstring...')
+
+            inst = Class()
+
+            # check that it is cached
+            self.assertEqual(get(inst), get(inst))
+
+            # check that it returns the right thing
+            self.assertEqual(get(inst)[0], 1)
+
+            # check that state isn't shared between instances
+            inst2 = Class()
+            self.assertNotEqual(get(inst), get(inst2))
+
+            # check that it behaves like a property when there's no instance
+            self.assertIsInstance(get(Class), cached_property)
+
+            # check that 'other_value' didn't become a property
+            self.assertTrue(callable(inst.other_value))
+
     def test_cached_property(self):
         """
         cached_property caches its value and that it behaves like a property
         """
-        class A:
+        class Class:
 
             @cached_property
             def value(self):
                 """Here is the docstring..."""
                 return 1, object()
 
+            @cached_property
+            def __foo__(self):
+                """Here is the docstring..."""
+                return 1, object()
+
             def other_value(self):
-                return 1
+                """Here is the docstring..."""
+                return 1, object()
 
             other = cached_property(other_value, name='other')
 
-        # docstring should be preserved
-        self.assertEqual(A.value.__doc__, "Here is the docstring...")
+        attrs = ['value', 'other', '__foo__']
+        for attr in attrs:
+            self.assertCachedPropertyWorks(attr, Class)
+
+    @unittest.skipUnless(PY36, '__set_name__ is new in Python 3.6')
+    def test_cached_property_auto_name(self):
+        """
+        cached_property caches its value and that it behaves like a property
+        on mangled methods or no name kwarg set.
+        """
+        class Class:
 
-        a = A()
```

Comment:
Ditto about the class `__name__` change.

Reply:
`__name__`?